File: data/derived_cot/rq3_prompting/results/translation/r1/cpp_to_py/code_output/TriCalculator.py
import math

class TriCalculator:
    PI = 3.14159265358979323846

    def __init__(self):
        pass

    def _round(self, value: float, precision: int) -> float:
        """Round half away from zero (like C++ std::round)."""
        factor = 10.0 ** precision
        scaled = value * factor
        # compute floor(|scaled| + 0.5) with sign
        if scaled >= 0:
            rounded = math.floor(scaled + 0.5)
        else:
            rounded = -math.floor(-scaled + 0.5)
        return rounded / factor

    def factorial(self, a: int) -> int:
        if a < 0:
            return 0
        if a == 0:
            return 1
        b = 1
        for i in range(1, a + 1):
            b *= i
        return b

    def taylor(self, x: float, n: int) -> float:
        a = 0.0
        x = x / 180.0 * self.PI
        for k in range(n):
            term = math.pow(x, 2 * k) / self.factorial(2 * k)
            if k % 2 == 0:
                a += term
            else:
                a -= term
        return a

    def cos(self, x: float) -> float:
        result = self.taylor(x, 17)
        print("Taylor result for cos: " + str(result))
        return self._round(result, 10)

    def sin(self, x: float) -> float:
        x = x / 180.0 * self.PI
        g = 0.0
        t = x
        n = 1
        while abs(t) >= 1e-15:
            g += t
            n += 1
            t = -t * x * x / ((2 * n - 1) * (2 * n - 2))
        return self._round(g, 10)

    def tan(self, x: float) -> float:
        cosine = self.cos(x)
        if abs(cosine) > 0.01:
            result = self.sin(x) / cosine
            return self._round(result, 10)
        else:
            return float('nan')

File: data/derived_cot/rq3_prompting/results/translation/glm-5.1/py_to_cpp/code_output/TriCalculator.cpp
#include <cmath>

class TriCalculator {
public:
    TriCalculator() {}

    double cos(double x) {
        return std::round(taylor(x, 50) * 1e10) / 1e10;
    }

    double factorial(int a) {
        double b = 1;
        while (a != 1) {
            b *= a;
            a -= 1;
        }
        return b;
    }

    double taylor(double x, int n) {
        double a = 1;
        x = x / 180 * M_PI;
        int count = 1;
        for (int k = 1; k < n; ++k) {
            if (count % 2 != 0) {
                a -= std::pow(x, 2 * k) / factorial(2 * k);
            } else {
                a += std::pow(x, 2 * k) / factorial(2 * k);
            }
            count += 1;
        }
        return a;
    }

    double sin(double x) {
        x = x / 180 * M_PI;
        double g = 0;
        double t = x;
        int n = 1;

        while (std::fabs(t) >= 1e-15) {
            g += t;
            n += 1;
            t = -t * x * x / (2 * n - 1) / (2 * n - 2);
        }
        return std::round(g * 1e10) / 1e10;
    }

    double tan(double x) {
        if (cos(x) != 0) {
            double result = sin(x) / cos(x);
            return std::round(result * 1e10) / 1e10;
        } else {
            return false; // false implicitly converts to 0.0, matching Python's False
        }
    }
};